Scripts/Python/get_device_list.py

```diff
@@ -34,61 +34,143 @@
-            device_response = requests.get(device_url, headers=headers, verify=False)
-            if device_response.status_code == 200:
-                json_data = device_response.json()
-                device_count = json_data['@odata.count']
-                if device_count > 0:
-                    if '@odata.nextLink' in json_data:
-                        next_link_url = base_uri + json_data['@odata.nextLink']
-                    while next_link_url:
-                        next_link_response = requests.get(next_link_url, headers=headers, verify=False)
-                        if next_link_response.status_code == 200:
-                            next_link_json_data = next_link_response.json()
-                            json_data['value'] += next_link_json_data['value']
-                            if '@odata.nextLink' in next_link_json_data:
-                                next_link_url = base_uri + next_link_json_data['@odata.nextLink']
-                            else:
-                                next_link_url = None
-                        else:
-                            print("Unable to retrieve device list from nextLink %s" % (next_link_url))
-                            next_link_url = None
-                    print("*** Device List ***")
-                    print(json.dumps(json_data, indent=4, sort_keys=True))
-                else:
-                    print("No devices managed by %s" % (ip_address))
-            else:
-                print("Unable to retrieve device list from %s" % (ip_address))
+            self.__headers['X-Auth-Token'] = session_info.headers['X-Auth-Token']
+        else:
+            print("Unable to create a session with appliance %s" % self.__session_input["ip"])
+
+    def __get_device_from_uri(self, uri):
+        json_data = {}
+        device_response = requests.get(uri, headers=self.__headers, verify=False)
+
+        if device_response.status_code == 200:
+            json_data = device_response.json()
         else:
-            print("Unable to create a session with appliance %s" % (ip_address))
-    except:
-        print ("Unexpected error:", sys.exc_info()[0])
+            print("Unable to retrieve device list from %s" % self.__session_input["ip"])
+
+        return json_data
+
+    def __get_device_list(self):
+        next_link_url = self.__base_uri + '/api/DeviceService/Devices'
+        self.json_data = None
+
+        while next_link_url is not None:
+            data = self.__get_device_from_uri(next_link_url)
+            next_link_url = None
+            if data['@odata.count'] <= 0:
+                print("No devices managed by %s" % self.__session_input["ip"])
+            if '@odata.nextLink' in data:
+                next_link_url = self.__base_uri + data['@odata.nextLink']
+            if self.json_data is None:
+                self.json_data = data
+            else:
+                self.json_data['value'] += data['value']
+
+    def __format_json(self):
```

Comment:
For consistency would it be easier to normalize the behavior of the json and csv methods - if there is no output path specified print the results to screen in the output format (csv.writer should be capable of writing to sys.stdout I'd think) - if a path is specified use the path and modify to unique (by addition of a counter or timestamp) -

Reply:
Have incorporated this change.